File: DesignPattenStudy/VisitorPattern/Java/Task.java
package VisitorPattern.Java;

import java.util.ArrayList;
import java.util.List;

public interface Task {

    public String doTask(Rider rider);
}

class DeliveryTask implements Task {

    @Override
    public String doTask(Rider rider) {
        return rider.doTask(this);
    }
}

class DriveVehicleTask implements Task {

    @Override
    public String doTask(Rider rider) {
        return rider.doTask(this);
    }
}

class RepairTask implements Task {

    @Override
    public String doTask(Rider rider) {
        return rider.doTask(this);
    }
}


class TotalRiderTask implements Task {

    ArrayList<Task> list = new ArrayList();

    TotalRiderTask() {
        list.add(new DeliveryTask());
        list.add(new DriveVehicleTask());
        list.add(new RepairTask());
    }

    @Override
    public String doTask(Rider rider) {

        String li = "";

        for (Task ri : list) {
            li += ri.doTask(rider) + ", ";

        }
        return li;
    }
}

File: DesignPattenStudy/VisitorPattern/Kotlin/Task.kt
package VisitorPattern.Kotlin

interface RiderTask {

    fun doTask(rider: Rider): String
}

class DeliveryTask : RiderTask {

    override fun doTask(rider: Rider): String = rider.doTask(this)
}

class DriveVehicleTask : RiderTask {

    override fun doTask(rider: Rider): String = rider.doTask(this)
}

class RepairVehicleTask : RiderTask {

    override fun doTask(rider: Rider): String = rider.doTask(this)
}



class TotalRiderTask : RiderTask {

    private val riderTaskList = mutableListOf<RiderTask>(DeliveryTask(), DeliveryTask(), RepairVehicleTask())

    override fun doTask(rider: Rider): String = riderTaskList.joinToString(separator = ", ") { it.doTask(rider) }
}

